examples/server/server.cpp

```diff
@@ -794,22 +739,16 @@ struct server_context {
                     continue;
                 }
 
-                // skip the slot if it does not contains prompt
-                if (!slot.prompt.is_string()) {
+                // skip the slot if it does not contains cached tokens
+                if (slot.prompt_tokens.empty()) {
                     continue;
                 }
 
-                // current slot's prompt
-                std::string slot_prompt = slot.prompt.get<std::string>();
-
-                // length of the current slot's prompt
-                int slot_prompt_len = slot_prompt.size();
-
                 // length of the Longest Common Prefix between the current slot's prompt and the input prompt
-                int lcp_len = longest_common_prefix(slot_prompt, prompt);
+                int lcp_len = longest_common_prefix(slot.cache_tokens, slot.prompt_tokens);
```

Comment:
OK sorry I see now. Indeed, I think the old code is quite confusing and does not really work (and that's why my new version is incorrect).

The old code is:

```cpp
std::string slot_prompt = slot.prompt.get<std::string>();
int lcp_len = longest_common_prefix(slot_prompt, prompt);
```

So when refactor I thought that `slot_prompt` is indeed `task.prompt`

The fix would be (as you said) to change the second argument to `task.prompt_tokens`. Do you want to make a PR @chrisstankevitz ?

```suggestion
                int lcp_len = longest_common_prefix(slot.cache_tokens, task.prompt_tokens);
```

Reply:
Also please note that here we compare:

`slot.cache_tokens, task.prompt_tokens`

But not

`slot.prompt_tokens, task.prompt_tokens`

Because some tokens stored in the last `prompt_tokens` may **not** be in cache